### PR TITLE
Batch RPC requests in fewer http requests automatically

### PR DESCRIPTION
**Pull request by bowd**
### Summary

In some cases when a client needs to handle a burst of >100 RPCs at once, it is advantageous to batch the requests in fewer HTTP calls by providing an array of RPCRequest payloads instead of a single RPCRequest object.

This is achieved by using a method inspired by [dataloaden](https://github.com/vektah/dataloaden) and [GraphQL dataloaders](https://gqlgen.com/reference/dataloaders/). It works transparently by batching multiple calls under the covers, so it doesn't affect the interface of `ethrpc` at all.

### Batching: how does it work

When batching is on, the HTTP provider uses a `dataloader` defined in `batch_loader.go`. The data loader acts as an arbiter for incoming requests. When it sees the first one in a cycle, it creates a batch and keeps appending requests to the batch until either the wait time is reached or the maximum batch size is reached, whichever comes first. At that point, the batch is closed and executed. During this time all function calls that initiated the request are blocked. 

Once the HTTP calls are done and the payloads decoded and passed back to the respective calls the functions return to where they were initiated. If errors happen at the request level, the error is passed down to each initiating call.

---

### Package layout

```
|-- httprpc
    |-- batch_loader.go - the batch loader
    |-- fetch.go - http request functions
    |-- httprpc.go - the base provider implementation
    `-- sync_loader.go - the sync loader
```

### Loaders

I cleaned up the code quite a bit and restructured it by introducing a new interface that the `HTTPProvider` depends on and that's the `RPCLoader`. It is responsible for how RPCs translate to HTTP requests. Currently, we have two loaders that implement this interface:

- `SyncLoader` which has the same behavior as before 1 RPC = 1 HTTP request
- `BatchLoader` which batches RPCs into HTTP requests based on the batch settings (max size and wait time).

When calling `httprpc.New` a `SyncLoader` is instantiated. This is in order to keep things backward compatible. In order to use batching you have to call `httprpc.NewWithLoader` and pass in an appropriate loader that you create before via `httprpc.NewSyncLoader` or  `httprpc.NewBatchLoader`. 

This approach is ok because it keeps `httprpc.New` working but it doesn't mirror how things work at the `ethrpc` level, where `New` actually takes an `RPCProvider` and `NewWithDefaults` makes a choice of what provider to use. If we want to keep the same pattern in the `httprpc` package we need a version bump. 

### Loaders interface

Currently, the `RPCLoader` interface has two methods:
- `Load` which takes an `*jsonrpc2.JSONRPCRequest` and returns a `[]byte`
- `Init` which takes the `*HTTPProvider`

The last thing I don't love about the code is this `Init` method. All the HTTP request code, HTTP client, etc exist in the `HTTPProvider`, and the loaders need to call those methods in order to actually execute the requests. `SyncLoader` uses `HTTPProvider.fetchSingle` and `BatchLoader` uses `HTTPProvider.fetchMultiple`.

If we want to really decouple things then we would have to implement another object called something like `RequestHandler` that contains all HTTP request related logic and these handlers are instantiated and managed in the loaders.

So basically `HTTPProvider` -> `RPCLoader` -> `RequestHandler` -> `HTTP request` instead of currently `HTTPProvider` -> `RPCLoader` -> `HTTPProvider` -> `HTTP request`.

### TODO

- [x] Proof of concept
- [x] Cleanup code
- [x] Implement configurable timeout
- [x] Ensure that the order of responses matches the order of requests, is this guaranteed?

### Questions

- [ ] Are we ok with the `New` + `NewWithLoader` interface?
- [ ] Are we ok with the `Init` func on `RPCLoader` that ties the loader to the HTTPProvider

### CODE REVIEW

**Comment by bowd** [comment]
@tzapu / @kwix share your thoughts when you get a chance. 

The code in `httprpc.go` is a bit messy right now, and there are a few things that are weird, most of all the fact that `CallRaw` isn't batched at the moment because there's no way to batch a call and not parse it to JSON before sending it back to the call point, because the raw response we get from the batch is an array of responses. So that's a bit of a weird point. And makes the code a bit messy right now and not as DRY as I would like it, but it's a POC, so it works.

I guess the question is if you agree with the general concept and how the `rpcloader` works. Its design is heavily borrowed from `dataloaden` minus some things like caching which didn't make sense in the context, so the code there is tried and tested.

Apart from that, I'd like to also add a configurable HTTP timeout but I don't know if it should be part of the initialization function or we should set a default and expose e separate setter. I'm just worried about ending up with too many params there.

File: ethrpc/provider/httprpc/httprpc.go [hunk 1]
```diff
@@ -8,14 +8,24 @@ import (
...
 	"strconv"
+	"time"
 
 	"github.com/alethio/web3-go/etherr"
 	"github.com/alethio/web3-go/jsonrpc2"
 )
 
+const (
+	// DefaultHTTPTimeout is the default timeout interval for http requests
+	DefaultHTTPTimeout = 3 * time.Second
```
Comment (tzapu) [inline ethrpc/provider/httprpc/httprpc.go:15]: should be customizable

**Review by tzapu — COMMENTED** [review]
i d go for separate setter as first option or go full blown config object

will look over the rest

**Comment by tzapu** [comment]
regarding the order of responses, i think it's taken for granted in the scraper  and  seems fine. @kwix  ?

**Comment by bowd** [comment]
I had a dream about how to refactor the code to be much nicer. Give me a few. Good to know about the order of responses, will make the code cleaner and make something possible.
I'm also thinking of a slightly different approach to decoding that would preserve the existing logic as much as possible and also parallelize JSON parsing: ~instead of unmarshaling the whole array of responses, I do regexp based split to isolate the objects `[{...}, {...}, {...}]` and then decode them individually~. I found out about `json.RawMessage` so I can decode as `[]json.RawMessage`, wicked! What I get is:
- We can also batch `CallRaw` requests
- I can keep more of the existing API and flow, that uses the `Raw` field on the `JSONRPCMessage`, otherwise, I couldn't populate it and that might introduce weirdness in the future.

**Comment by bowd** [comment]
Done a bit of rework on this and updated the description to add some more context 🏗Basically I only have those two questions at the end that need to be answered to push this forward.

**Comment by lacasian** [comment]
> regarding the order of responses, i think it's taken for granted in the scraper and seems fine. @kwix ?

yep, takes it for granted but I have no idea if this is a standard or not ... for me it feels a bit wrong to do that but haven't seen any issues so far

**Review by tzapu — APPROVED** [review]
i wonder why tests aren't running

**Comment by bowd** [comment]
Reopened as #5, to see if CI runs
